File: app/src/test/java/br/eng/ecarrara/vilibra/data/cache/BookContentProviderCacheTest.java
package br.eng.ecarrara.vilibra.data.cache;

import android.content.Context;
import android.database.Cursor;

import org.junit.After;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricGradleTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import java.security.InvalidParameterException;

import br.eng.ecarrara.vilibra.BuildConfig;
import br.eng.ecarrara.vilibra.book.data.datasource.contentprovider.BookContentProviderCache;
import br.eng.ecarrara.vilibra.core.data.datasource.contentprovider.VilibraContract;
import br.eng.ecarrara.vilibra.core.data.RxCache;
import br.eng.ecarrara.vilibra.book.domain.entity.Book;
import br.eng.ecarrara.vilibra.fixture.BookFixture;
import br.eng.ecarrara.vilibra.fixture.VilibraProviderFixture;
import br.eng.ecarrara.vilibra.utils.RobolectricUtils;

import static org.hamcrest.core.Is.is;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.junit.Assert.assertThat;

@RunWith(RobolectricGradleTestRunner.class)
@Config(constants = BuildConfig.class)
public class BookContentProviderCacheTest {

    private Context context;
    private VilibraProviderFixture vilibraProviderFixture;

    private Book testBook;
    private RxCache<String, Book> bookContentProviderCache;

    @BeforeClass public static void setupRobolectric() {
        RobolectricUtils.redirectLogsToSystemOutput();
        RobolectricUtils.setupVilibraContentProvider();
    }

    @Before public void prepareTest() {
        this.context = RuntimeEnvironment.application;
        this.vilibraProviderFixture = new VilibraProviderFixture(this.context);
        this.vilibraProviderFixture.prepareTestProvider();
        this.testBook = this.vilibraProviderFixture.getProAndroid4();

        this.bookContentProviderCache = new BookContentProviderCache(this.context);
    }

    @After public void cleanUp() {
        this.vilibraProviderFixture.clearVilibraDatabase();
    }

    @Test(expected = InvalidParameterException.class)
    public void testNullInitializationThrowsException() {
        this.bookContentProviderCache = new BookContentProviderCache(null);
    }

    @Test public void testGetWithIsbn10() throws Exception {
        bookContentProviderCache
                .get(this.testBook.getIsbn10())
                .test()
                .assertComplete()
                .assertValue(testBook);
    }

    @Test public void testGetWithIsbn13() throws Exception {
        bookContentProviderCache
                .get(this.testBook.getIsbn13())
                .test()
                .assertComplete()
                .assertValue(testBook);
    }

    @Test public void testPutWithIsbn10() throws Exception {
        this.vilibraProviderFixture.clearVilibraDatabase();
        Book anotherBook = BookFixture.INSTANCE.getTestBookDevsTestBook();

        bookContentProviderCache
                .put(anotherBook.getIsbn10(), anotherBook)
                .andThen(bookContentProviderCache.get(anotherBook.getIsbn10()))
                .test()
                .assertComplete()
                .assertValue(anotherBook);
    }

    @Test public void testPutWithIsbn13() throws Exception {
        this.vilibraProviderFixture.clearVilibraDatabase();
        Book anotherBook = BookFixture.INSTANCE.getTestBookDevsTestBook();

        bookContentProviderCache
                .put(anotherBook.getIsbn13(), anotherBook)
                .andThen(bookContentProviderCache.get(anotherBook.getIsbn13()))
                .test()
                .assertComplete()
                .assertValue(anotherBook);
    }

    @Test public void testPutWithIsbn10AdGetWithIsbn13() throws Exception {
        this.vilibraProviderFixture.clearVilibraDatabase();
        Book anotherBook = BookFixture.INSTANCE.getTestBookDevsTestBook();

        bookContentProviderCache
                .put(anotherBook.getIsbn10(), anotherBook)
                .andThen(bookContentProviderCache.get(anotherBook.getIsbn13()))
                .test()
                .assertComplete()
                .assertValue(anotherBook);
    }

    @Test public void testPutWithIsbn13AdGetWithIsbn10() throws Exception {
        this.vilibraProviderFixture.clearVilibraDatabase();
        Book anotherBook = BookFixture.INSTANCE.getTestBookDevsTestBook();

        bookContentProviderCache
                .put(anotherBook.getIsbn13(), anotherBook)
                .andThen(this.bookContentProviderCache.get(anotherBook.getIsbn10()))
                .test()
                .assertValue(anotherBook);
    }

    @Test public void testRemoveWithIsbn10() throws Exception {
        bookContentProviderCache
                .remove(this.testBook.getIsbn10())
                .andThen(bookContentProviderCache.get(this.testBook.getIsbn10()))
                .test()
                .assertComplete()
                .assertValue(Book.Companion.getNO_BOOK());
    }

    @Test public void testRemoveWithIsbn13() throws Exception {
        bookContentProviderCache
                .remove(this.testBook.getIsbn13())
                .andThen(bookContentProviderCache.get(this.testBook.getIsbn10()))
                .test()
                .assertComplete()
                .assertValue(Book.Companion.getNO_BOOK());
    }

    @Test public void testIsCachedWithIsbn10() throws Exception {
        bookContentProviderCache
                .isCached(this.testBook.getIsbn10())
                .test()
                .assertComplete()
                .assertValue(true);
    }

    @Test public void testIsCachedWithIsbn13() throws Exception {
        bookContentProviderCache
                .isCached(this.testBook.getIsbn13())
                .test()
                .assertComplete()
                .assertValue(true);
    }

    @Test public void testClear() throws Exception {
        bookContentProviderCache
                .clear()
                .test()
                .assertComplete();

        Cursor retrievedBooks = this.context.getContentResolver()
                .query(
                        VilibraContract.BookEntry.CONTENT_URI,
                        new String[]{VilibraContract.BookEntry.COLUMN_BOOK_ID},
                        null, null, null);

        boolean areThereAnyBooksInTheCache = retrievedBooks.moveToFirst();
        assertThat(areThereAnyBooksInTheCache, is(false));
    }
}

File: app/src/test/kotlin/br/eng/ecarrara/vilibra/book/data/datasource/contentprovider/BookContentProviderCacheTest.kt
package br.eng.ecarrara.vilibra.book.data.datasource.contentprovider

import android.content.Context
import br.eng.ecarrara.vilibra.BuildConfig
import br.eng.ecarrara.vilibra.book.data.datasource.BookLocalCache
import br.eng.ecarrara.vilibra.book.domain.entity.Book
import br.eng.ecarrara.vilibra.data.VilibraContract
import br.eng.ecarrara.vilibra.data.VilibraProvider
import br.eng.ecarrara.vilibra.fakedata.BookFakeDataFactory
import br.eng.ecarrara.vilibra.fakedata.VilibraProviderFakeDataInitializer
import br.eng.ecarrara.vilibra.utils.RobolectricUtils
import org.hamcrest.core.Is.`is`
import org.junit.After
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowContentResolver

@RunWith(RobolectricTestRunner::class)
@Config(constants = BuildConfig::class)
class BookContentProviderCacheTest {

    private lateinit var context: Context

    private lateinit var vilibraProviderFixture: VilibraProviderFakeDataInitializer
    private lateinit var testBook: Book

    private lateinit var bookLocalCache: BookLocalCache

    @Before
    fun prepareTest() {
        this.context = RuntimeEnvironment.application!!

        ShadowContentResolver.registerProviderInternal(
                VilibraContract.CONTENT_AUTHORITY,
                Robolectric.setupContentProvider(VilibraProvider::class.java)
        )

        this.vilibraProviderFixture = VilibraProviderFakeDataInitializer(this.context)
        this.vilibraProviderFixture.prepareTestProvider()
        this.testBook = this.vilibraProviderFixture.getProAndroid4()!!

        this.bookLocalCache = BookContentProviderCache(this.context)
    }

    @After
    fun cleanUp() {
        this.vilibraProviderFixture.clearContentProviderData()
    }

    @Test
    @Throws(Exception::class)
    fun testGetWithIsbn10() {
        bookLocalCache.get(testBook.isbn10)
                .test()
                .assertValue(testBook)
                .assertComplete()
    }

    @Test
    @Throws(Exception::class)
    fun testGetWithIsbn13() {
        bookLocalCache.get(testBook.isbn13)
                .test()
                .assertValue(testBook)
                .assertComplete()
    }

    @Test
    @Throws(Exception::class)
    fun testPutWithIsbn10() {

        vilibraProviderFixture.clearContentProviderData()
        val anotherBook = BookFakeDataFactory.fakeBookDevsTest
        bookLocalCache.put(anotherBook.isbn10, anotherBook)
                .test()
                .assertComplete()

        bookLocalCache.get(anotherBook.isbn10)
                .test()
                .assertValue(anotherBook)
                .assertComplete()
    }

    @Test
    @Throws(Exception::class)
    fun testPutWithIsbn13() {
        vilibraProviderFixture.clearContentProviderData()
        val anotherBook = BookFakeDataFactory.fakeBookDevsTest

        bookLocalCache.put(anotherBook.isbn13, anotherBook)
                .test()
                .assertComplete()

        bookLocalCache.get(anotherBook.isbn13)
                .test()
                .assertValue(anotherBook)
                .assertComplete()
    }

    @Test
    @Throws(Exception::class)
    fun testPutWithIsbn10AdGetWithIsbn13() {
        vilibraProviderFixture.clearContentProviderData()
        val anotherBook = BookFakeDataFactory.fakeBookDevsTest

        bookLocalCache.put(anotherBook.isbn10, anotherBook)
                .test()
                .assertComplete()

        bookLocalCache.get(anotherBook.isbn13)
                .test()
                .assertValue(anotherBook)
                .assertComplete()
    }

    @Test
    @Throws(Exception::class)
    fun testPutWithIsbn13AndGetWithIsbn10() {
        vilibraProviderFixture.clearContentProviderData()
        val anotherBook = BookFakeDataFactory.fakeBookDevsTest

        bookLocalCache.put(anotherBook.isbn13, anotherBook)
                .test()
                .assertComplete()

        bookLocalCache.get(anotherBook.isbn10)
                .test()
                .assertValue(anotherBook)
                .assertComplete()
    }

    @Test
    @Throws(Exception::class)
    fun testRemoveWithIsbn10() {
        bookLocalCache.remove(this.testBook.isbn10)
                .test()
                .assertComplete()

        bookLocalCache.get(this.testBook.isbn10)
                .test()
                .assertValue(Book.NO_BOOK)
                .assertComplete()
    }

    @Test
    @Throws(Exception::class)
    fun testRemoveWithIsbn13() {
        bookLocalCache.remove(this.testBook.isbn13)
                .test()
                .assertComplete()

        bookLocalCache.get(this.testBook.isbn10)
                .test()
                .assertValue(Book.NO_BOOK)
                .assertComplete()
    }

    @Test
    @Throws(Exception::class)
    fun testIsCachedWithIsbn10() {
        bookLocalCache.isCached(testBook.isbn10)
                .test()
                .assertValue(true)
                .assertComplete()
    }

    @Test
    @Throws(Exception::class)
    fun testIsCachedWithIsbn13() {
        bookLocalCache.isCached(testBook.isbn13)
                .test()
                .assertValue(true)
                .assertComplete()
    }

    @Test
    @Throws(Exception::class)
    fun testClear() {

        bookLocalCache.clear()
                .test()
                .assertComplete()

        val retrievedBooks = context.contentResolver.query(
                VilibraContract.BookEntry.CONTENT_URI,
                arrayOf(VilibraContract.BookEntry.COLUMN_BOOK_ID),
                null, null, null
        )

        val areThereAnyBooksInTheCache = retrievedBooks!!.moveToFirst()
        assertThat(areThereAnyBooksInTheCache, `is`(false))

        retrievedBooks.close()
    }

    companion object {
        @JvmStatic
        @BeforeClass
        fun setupRobolectric() {
            RobolectricUtils.redirectLogsToSystemOutput()
        }
    }
}
